test/unittests/execute_floating_point_test.cpp

```diff
@@ -440,6 +492,72 @@ TYPED_TEST(execute_floating_point_types, neg)
     }
 }
 
+TYPED_TEST(execute_floating_point_types, ceil)
+{
+    auto instance = instantiate(parse(this->get_unop_code(Instr::f32_ceil)));
+    const auto exec = [&](auto arg) { return execute(*instance, 0, {arg}); };
+
+    // fceil(-q) = -0  (if -1 < -q < 0)
+    EXPECT_THAT(exec(std::nextafter(-TypeParam{1}, TypeParam{0})), Result(-TypeParam{0}));
+    EXPECT_THAT(exec(-FP<TypeParam>::Limits::denorm_min()), Result(-TypeParam{0}));
+
+    for (const auto& [arg, expected_trunc] :
+        execute_floating_point_types<TypeParam>::positive_trunc_tests)
+    {
+        // For positive values, the ceil() is trunc() + 1, unless the input is already an integer.
+        const auto expected_pos =
+            (arg == expected_trunc) ? expected_trunc : expected_trunc + TypeParam{1};
+        EXPECT_THAT(exec(arg), Result(expected_pos)) << arg << ": " << expected_pos;
+
+        // For negative values, the ceil() is trunc().
+        EXPECT_THAT(exec(-arg), Result(-expected_trunc)) << -arg << ": " << -expected_trunc;
+    }
+}
+
+TYPED_TEST(execute_floating_point_types, floor)
+{
+    auto instance = instantiate(parse(this->get_unop_code(Instr::f32_floor)));
+    const auto exec = [&](auto arg) { return execute(*instance, 0, {arg}); };
+
+    // ffloor(+q) = +0  (if 0 < +q < 1)
+    EXPECT_THAT(exec(FP<TypeParam>::Limits::denorm_min()), Result(TypeParam{0}));
+    EXPECT_THAT(exec(std::nextafter(TypeParam{1}, TypeParam{0})), Result(TypeParam{0}));
+
+    for (const auto& [arg, expected_trunc] :
+        execute_floating_point_types<TypeParam>::positive_trunc_tests)
+    {
+        // For positive values, the floor() is trunc().
+        EXPECT_THAT(exec(arg), Result(expected_trunc)) << arg << ": " << expected_trunc;
+
+        // For negative values, the floor() is trunc() - 1, unless the input is already an integer.
+        const auto expected_neg =
+            (arg == expected_trunc) ? -expected_trunc : -expected_trunc - TypeParam{1};
+        EXPECT_THAT(exec(-arg), Result(expected_neg)) << -arg << ": " << expected_neg;
+    }
+}
+
+TYPED_TEST(execute_floating_point_types, trunc)
+{
+    auto instance = instantiate(parse(this->get_unop_code(Instr::f32_trunc)));
+    const auto exec = [&](auto arg) { return execute(*instance, 0, {arg}); };
+
+    // ftrunc(+q) = +0  (if 0 < +q < 1)
+    EXPECT_THAT(exec(FP<TypeParam>::Limits::denorm_min()), Result(TypeParam{0}));
+    EXPECT_THAT(exec(std::nextafter(TypeParam{1}, TypeParam{0})), Result(TypeParam{0}));
```

Comment:
Should these have swapped operand: `nextafter(0, 1)` e.g. the smallest number larger than 0 ?

Reply:
The `denorm_min` is the smallest > 0.
The `nextafter(0, 1)` is the largest < 1.

Reply:
If `nextafter(0, 1)` is the largest < 1, then why do you have `nextafter(0, 1)` here?

Reply:
Wrong comment.
`nextafter(0,1)` is `denorm_min` - i.e. it is the next after 0 towards 1.
`nextafter(1,0` is the largest < 1.